nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java

```diff
@@ -307,6 +313,41 @@ private UpdateableReference loadReference(String requestedRef) {
     }
   }
 
+  private UpdateableReference loadReferenceAtTime(String requestedRef, Instant timestamp) {
+    try {
+      Reference ref = requestedRef == null ? client.getTreeApi().getDefaultBranch()
+          : client.getTreeApi().getReferenceByName(requestedRef);
+      if (ref instanceof Hash) {
+        LOGGER.warn("Cannot specify a hash {} and timestamp {} together. " +
+            "The timestamp is redundant and has been ignored", requestedRef, timestamp);
+        return new UpdateableReference(ref, client.getTreeApi());
+      }
+      // NOTE fetching logs is currently potentially expensive
+      // todo this algorithm doesn't currently handle assign operations. The resulting hash may
+      // be the hash from the branch that requestedRef was reassigned to at time 't' rather than the hash
+      // of this ref at time 't' BEFORE reassignment. This logic needs to move to the server side and will
+      // be done in a follow up commit.
+      List<CommitMeta> ops = client.getTreeApi().getCommitLog(ref.getName()).getOperations();
+      for (CommitMeta info : ops) {
+        if (info.getCommitTime() != null && Instant.ofEpochMilli(info.getCommitTime()).isBefore(timestamp)) {
+          return new UpdateableReference(ImmutableHash.builder().name(info.getHash()).build(), client.getTreeApi());
```

Comment:
Won't this return the first reference? Is there a guarantee that the commit log is in reverse order?

Reply:
The commit log will always be in reverse order. (similar to how git would show its log). So this will show most recent commit first and base commit last. I have added another test to verify the behaviour.